verilog/tools/kythe/kythe_facts_extractor.h

```diff
@@ -0,0 +1,156 @@
+
+#include <string>
+#include <utility>
+
+#include "verilog/tools/kythe/indexing_facts_tree.h"
+#include "verilog/tools/kythe/indexing_facts_tree_context.h"
+#include "verilog/tools/kythe/indexing_facts_tree_extractor.h"
+#include "verilog/tools/kythe/kythe_facts.h"
+
+namespace verilog {
+namespace kythe {
+
+// Streamable printing class for kythe facts.
+// Usage: stream << KytheFactsPrinter(*tree_root);
+class KytheFactsPrinter {
+ public:
+  explicit KytheFactsPrinter(const IndexingFactNode& root) : root_(root) {}
+
+  std::ostream& Print(std::ostream&) const;
+
+ private:
+  // The root of the indexing facts tree to extract kythe facts from.
+  const IndexingFactNode& root_;
+};
+
+// Responsible for traversing IndexingFactsTree and processing its different
+// nodes to produce kythe indexing facts.
+class KytheFactsExtractor {
+ public:
+  explicit KytheFactsExtractor(absl::string_view file_path,
+                               std::ostream* stream)
+      : file_path_(file_path), stream_(stream) {}
+
+  void Visit(const IndexingFactNode&);
+
+ private:
+  // Container with a stack of VNames to hold context of VNames during traversal
+  // of an IndexingFactsTree.
+  class VNameContext : public verible::AutoPopStack<const VName*> {
+   public:
+    typedef verible::AutoPopStack<const VName*> base_type;
+
+    // member class to handle push and pop of stack safely
+    using AutoPop = base_type::AutoPop;
+
+    // returns the top VName of the stack
+    const VName& top() const { return *ABSL_DIE_IF_NULL(base_type::top()); }
+  };
+
+  // Container with a stack of Scopes to hold the accessible scopes during
+  // traversing an Indexing Facts Tree.
+  // This is used to get the definitions of some reference.
+  class ScopeContext : public verible::AutoPopStack<std::vector<VName>*> {
+   public:
+    typedef verible::AutoPopStack<std::vector<VName>*> base_type;
+
+    // member class to handle push and pop of stack safely
+    using AutoPop = base_type::AutoPop;
+
+    // returns the top VName of the stack
+    std::vector<VName>& top() { return *ABSL_DIE_IF_NULL(base_type::top()); }
+
+    // Search function to get the VName of a definitions of some reference.
+    // It loops over the scopes in reverse order and loops over every scope in
+    // reverse order to find a definition for the variable with given prefix
+    // signature.
+    // e.g
+    // {
+    //    bar#module,
+    //    foo#module,
+    // }
+    // {
+    //    other scope,
+    // }
+    // Given bar#module it return the whole VName of that definition.
+    // And if more than one match is found the first would be returned.
+    const VName* SearchForDefinition(std::string prefix) {
+      for (const auto& scope : verible::make_range(rbegin(), rend())) {
+        for (const VName& vname :
+             verible::make_range(scope->rbegin(), scope->rend())) {
+          if (vname.signature.find(prefix) != std::string::npos) {
```

Comment:
The intention is to make the signatures of some definition unique in a way that won't collide with other definitions.
Till now I don't think it will math unintentional substrings.

Reply:
What if prefix is a short string like `a`?  will this match on signatures containing `a`?

Reply:
The prefix shouldn't be `a` it should be something more specific like `a#variable` or `a#module`.

Reply:
We clarified in discussion that `signature` is a concatenation of name-type tuples, of arbitrary depth.
We may consider adding structure to `signature` later.
[no change for now]